erpnext/accounts/doctype/payment_manager/payment_manager.py

```diff
@@ -0,0 +1,429 @@
+				'party_type': payment.party_type,
+				'party': frappe.get_value("Payment Entry", payment.reference_name, "party"),
+				'unadjusted_amount' : float(amount),
+				'allocated_amount' : min(outstanding_amount, amount)
+			}))
+			amount -= outstanding_amount
+		if lst:
+			from erpnext.accounts.utils import reconcile_against_document
+			try:
+				reconcile_against_document(lst)
+			except:
+				frappe.throw("Exception occurred while reconciling {0}".format(payment.reference_name))
+
+	def submit_payment_entries(self):
+		for payment in self.new_transaction_items:
+			if payment.reference_name is None: continue
+			doc = frappe.get_doc(payment.reference_type, payment.reference_name)
+			if doc.docstatus == 1:
+				if (payment.reference_type == "Journal Entry"): continue
+				if doc.unallocated_amount == 0: continue
+				print("Reconciling payment {0}".format(payment.reference_name))
+				self.update_payment_entry(payment)
+			else:
+				print("Submitting payment {0}".format(payment.reference_name))
+				if (payment.reference_type == "Payment Entry"):
+					if (payment.payment_reference):
+						doc.reference_no = payment.payment_reference
+					doc.mode_of_payment = payment.mode_of_payment
+				doc.save()
+				doc.submit()
+		self.move_reconciled_entries()
+		self.populate_matching_invoices()
+
+	def move_reconciled_entries(self):
+		idx = 0
+		while idx < len(self.new_transaction_items):
+			entry = self.new_transaction_items[idx]
+			print("Checking transaction {0}: {2} in {1} entries".format(idx, len(self.new_transaction_items), entry.description))
+			idx += 1
+			if entry.reference_name is None: continue
+			doc = frappe.get_doc(entry.reference_type, entry.reference_name)
+			if doc.docstatus == 1 and (entry.reference_type == "Journal Entry" or doc.unallocated_amount == 0):
+				self.remove(entry)
+				rc_entry = self.append('reconciled_transaction_items', {})
+				dentry = entry.as_dict()
+				dentry.pop('idx', None)
+				rc_entry.update(dentry)
+				idx -= 1
+
+
+def get_matching_journal_entries(from_date, to_date, account, against, amount):
+	query = """select voucher_no, posting_date, account, against, debit_in_account_currency as debit, credit_in_account_currency as credit
+							      from `tabGL Entry`
+								  where posting_date between '{0}' and '{1}' and account = '{2}' and against = '{3}' and debit = '{4}'
+								  """.format(from_date, to_date, account, against, amount)
+	jv_entries = frappe.db.sql(query, as_dict=True)
+	#print("voucher query:{0}\n Returned {1} entries".format(query, len(jv_entries)))
+	return jv_entries
+
+def get_payments_matching_invoice(invoice, amount, pay_date):
+	query = """select pe.name as reference_name, per.reference_doctype as reference_type, per.outstanding_amount, per.allocated_amount
+				from `tabPayment Entry Reference` as per JOIN `tabPayment Entry` as pe on pe.name = per.parent
+				where per.reference_name='{0}' and (posting_date='{1}' or reference_date='{1}') and pe.docstatus != 2
```

Comment:
this will cause SQL injection.

Use `sql('query', values)` syntax to avoid injections

Reply:
Ah, I was wondering about the value of that particular syntax all along. I will work on that.